src/tree/index.tsx

```diff
@@ -0,0 +1,422 @@
+					selectable && selectNode(activeNode);
+					checkable && checkNode(activeNode, !checkedNodes.includes(activeNode));
+				}
+				break;
+
+			// next
+			case Keys.Down:
+				event.preventDefault();
+				activateNode(nodes[(activeIndex + 1) % nodes.length].id);
+				break;
+
+			// previous
+			case Keys.Up:
+				event.preventDefault();
+				if (activeIndex - 1 < 0) {
+					activateNode(nodes[nodes.length - 1].id);
+				} else {
+					activateNode(nodes[activeIndex - 1].id);
+				}
+				break;
+
+			// expand
+			case Keys.Right:
+				event.preventDefault();
+				if (
+					activeNode &&
+					!activeExpandedNodes.includes(activeNode) &&
+					nodes[activeIndex].hasChildren
+				) {
+					expandNode(activeNode);
+				}
+				break;
+
+			// collapse
+			case Keys.Left:
+				event.preventDefault();
+				if (activeNode && activeExpandedNodes.includes(activeNode)) {
+					collapseNode(activeNode);
+				}
+				break;
+		}
+	}
+
+	function mapNodeTree(nodeId: string = 'root') {
+		const options = createOptions(nodeId);
+		const info = meta(template, options({ query: { parent: nodeId } }), true);
+
+		if (info === undefined || isLoading(template, options())) {
+			return <LoadingIndicator />;
+		}
+
+		const results = getOrRead(
+			template,
+			options({ query: { parent: nodeId }, size: info.total })
+		);
+
+		const nodes = flat(results);
+
+		return (
+			<ol
+				classes={[
+					nodeId === 'root' ? classes.root : null,
+					classes.nodeParent,
+					theme.variant()
+				]}
+				focus={() => shouldFocus}
+				onkeydown={onKeyDown}
+				tabIndex={0}
+			>
+				{nodes.map((node) => {
+					const isExpanded = controlledExpandedNodes
+						? controlledExpandedNodes.indexOf(node.id) !== -1
+						: expandedNodes.indexOf(node.id) !== -1;
+					return (
+						<li
+							classes={[
+								classes.node,
+								node.hasChildren && classes.leaf,
+								selectable && classes.selectable,
+								node.id === selectedNode && classes.selected
+							]}
+						>
+							<TreeNode
+								activeNode={activeNode}
+								checkable={checkable}
+								selectable={selectable}
+								checkedNodes={controlledCheckedNodes || checkedNodes}
+								selectedNode={selectedNode}
+								disabledNodes={disabledNodes || []}
+								expandedNodes={controlledExpandedNodes || expandedNodes}
+								parentSelection={parentSelection}
+								node={node}
+								onActive={activateNode}
+								onSelect={selectNode}
+								onCheck={checkNode}
+								onExpand={(n, expanded) => {
+									if (expanded) {
+										expandNode(n);
+									} else {
+										collapseNode(n);
+									}
+								}}
+							>
+								{itemRenderer || defaultRenderer}
+							</TreeNode>
+							{isExpanded && mapNodeTree(node.id)}
+						</li>
+					);
+				})}
+			</ol>
+		);
+	}
+	return mapNodeTree();
+});
+
+interface TreeNodeProperties {
+	checkable: boolean;
+	selectable: boolean;
+	activeNode?: string;
+	selectedNode?: string;
+	checkedNodes: string[];
```

Comment:
i don't think that a tree node should know anything bout the selected nodes, active nodes etc, it should simply be passed a boolean if it's active / selected / checked etc. Please correct me if I am wrong.

Reply:
yup I agree here, it seems the only reason these values are being passed is to calculate `selected` `checked` etc